backend/consensus/base.py

```diff
@@ -202,20 +218,44 @@ async def exec_transaction(
 
             num_validators = len(remaining_validators) + 1
 
-            contract_snapshot = contract_snapshot_factory(transaction.to_address)
+            contract_snapshot_supplier = lambda: contract_snapshot_factory(
+                transaction.to_address
+            )
```

Comment:
is this needed because you need to get the update for the validators?
If that's the case probably this means that we should do these changes in the ContractSnapshot class

Reply:
Because if you pass the single snapshot to all validators (which update it during execution) something unexpected happens. Previously changes were done in the end, but now `ContractSnapshot` is mutable

Also, how contract snapshot state operations are implemented is extremely unoptimized, but I don't really know how to fix it without messing up the database, so I left it for future patches

Reply:
Just to be sure that I understand. With this change, the leader and all the validators takes a new instance of the contract snapshot?

Reply:
Yes. Previously they got the same for an executed contract and (I believe) different for other contracts (due to factory). Now they don't share any contract state between each other